L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorAutoEncoderImpl.cc

```diff
@@ -0,0 +1,229 @@
+
+    // create a new session and add the graphDef
+    session_decoder_.push_back(std::unique_ptr<tensorflow::Session>{tensorflow::createSession(graphDef_decoder_.get())});
+
+    //extract encoder tenser names from first graph, check that rest of the names are consistent
+    if (modelFilePset == modelFilePaths_.front()){
+      inputTensorName_encoder_ = graphDef_encoder_.get()->node(0).name();
+      outputTensorName_encoder_ = graphDef_encoder_.get()->node(graphDef_encoder_.get()->node_size()-1).name();
+      inputTensorName_decoder_ = graphDef_decoder_.get()->node(0).name();
+      outputTensorName_decoder_ = graphDef_decoder_.get()->node(graphDef_decoder_.get()->node_size()-1).name();
+    } else {
+      if (inputTensorName_encoder_ != graphDef_encoder_.get()->node(0).name()){
+	throw cms::Exception("BadInitialization") << "provided list of encoder graphs have different input nodes";
+      }
+      if (outputTensorName_encoder_ != graphDef_encoder_.get()->node(graphDef_encoder_.get()->node_size()-1).name()){
+	throw cms::Exception("BadInitialization") << "provided list of encoder graphs have different output nodes";
+      }
+      if (inputTensorName_decoder_ != graphDef_decoder_.get()->node(0).name()){
+	throw cms::Exception("BadInitialization") << "provided list of decoder graphs have different input nodes";
+      }
+      if (outputTensorName_decoder_ != graphDef_decoder_.get()->node(graphDef_decoder_.get()->node_size()-1).name()){
+	throw cms::Exception("BadInitialization") << "provided list of decoder graphs have different output nodes";
+      }
+    }
+  }
+
+  // check that the appropriate number of links have been specified
+  if (linkToGraphMap_.size() <= maxNumberOfLinks_){
+    throw cms::Exception("BadInitialization") << "Autoencoder graph number must be specified for all link allocation possibilities. Only " << linkToGraphMap_.size() << " values specified while " << maxNumberOfLinks_ << "links are possible";
+  }
+
+  // check that all graph indices specified exist in the model file lists
+  for (const auto& graphNumber: linkToGraphMap_) {
+    if (graphNumber >= modelFilePaths_.size()){
+      throw cms::Exception("BadInitialization") << "Autoencoder graph number  "<< graphNumber << " is larger than the size of the provided list of graphs " << modelFilePaths_.size();	
+    }
+  }
+}
+
+void HGCalConcentratorAutoEncoderImpl::select(unsigned nLinks,
+                                              const std::vector<l1t::HGCalTriggerCell>& trigCellVecInput,
+                                              std::vector<l1t::HGCalTriggerCell>& trigCellVecOutput,
+                                              std::vector<l1t::HGCalConcentratorData>& ae_encodedLayer_Output) {
+
+  //reset intputs to 0 to account for zero suppressed trigger cells
+  mipPt_.fill(0);
+  uncompressedCharge_.fill(0);
+  compressedCharge_.fill(0);
+
+  double modSum = 0;
+
+  int bitsPerOutput = outputBitsPerLink_.at(nLinks);
```

Comment:
I had been setting -1 for the number of bits (or any negative value really) to keep the full precision of the outputs instead of truncating so that we can test the impact of the output quantization.  I now see though that this will cause a problem in the bit shifted maxInputSize variables, so have added a check to only do the bit shift for positive values.

```
  double inputMaxIntSize = 1;
  if (bitsPerInput_ > 0)
    inputMaxIntSize = 1 << bitsPerInput_;
  double outputMaxIntSize = 1;
  if (bitsPerOutput > 0)
    outputMaxIntSize = 1 << bitsPerOutput;
  double outputMaxIntSizeGlobal = 1;
  if (maxBitsPerOutput_ > 0)
    outputMaxIntSizeGlobal = 1 << maxBitsPerOutput_;
```